raft/raft.go

```diff
@@ -782,9 +782,11 @@ func (r *raft) Step(m pb.Message) error {
 	case pb.MsgVote, pb.MsgPreVote:
 		// The m.Term > r.Term clause is for MsgPreVote. For MsgVote m.Term should
 		// always equal r.Term.
-		if (r.Vote == None || m.Term > r.Term || r.Vote == m.From) && r.raftLog.isUpToDate(m.Index, m.LogTerm) {
-			r.logger.Infof("%x [logterm: %d, index: %d, vote: %x] cast %s for %x [logterm: %d, index: %d] at term %d",
-				r.id, r.raftLog.lastTerm(), r.raftLog.lastIndex(), r.Vote, m.Type, m.From, m.LogTerm, m.Index, r.Term)
+		if m.Type == pb.MsgVote && m.Term != r.Term {
+			panic(fmt.Sprintf("m.Type == pb.MsgVote && m.Term != r.Term: %d != %d", m.Term, r.Term))
+		}
+		if (r.Vote == None || r.Vote == m.From || (m.Type == pb.MsgPreVote && m.Term > r.Term)) &&
```

Comment:
seems no different to previous.   
do you mean this?
 `(r.Vote == None || r.Vote == m.From) && (m.Type == pb.MsgPreVote || (m.Type == pb.MsgPreVote && m.Term > r.Term)) &&`

Reply:
the difference that we do not grant prevote if m.term <= t.term like before.

Reply:
the if checking seems incorrect though. i am fixing it right now.

Reply:
I'm looking at this again and the change doesn't make sense to me. It doesn't change the behavior of MsgPreVote - for MsgPreVote the condition was `m.Term > r.Term` both before and after the change. And adding the guard `m.Type == pb.MsgPreVote` doesn't really change anything because we know that for MsgVote, the terms are equal and the `m.Term > r.Term` condition will never be true. 

The new code is slightly more explicit, so we might want to merge this as a cleanup, but I don't think there's a bug being fixed here. I've been unable to modify the test in #8571 to find any cases where this change makes a difference. 

I do see one small bug: If `m.Type == MsgPreVote && m.Term == r.Term && r.Vote == None && r.lead != None`, then we will grant a PreVote when we shouldn't. However, there are not going to be enough nodes with `r.Vote == None && r.lead != None` to allow the PreVote to reach a quorum unless there have also been membership changes during the term. 

I'll send a PR with some cleanups here. Testing the "leader without vote" scenario is difficult because it relies on membership changes, so unless you feel strongly I think I'll skip the test for now (I'm going to do some work on PreVote testing separately).